nameserver/dns.go

```diff
@@ -27,15 +29,19 @@ const (
 )
 
 type DNSServer struct {
-	ns      *Nameserver
-	domain  string
-	ttl     uint32
-	address string
-
-	servers   []*dns.Server
-	upstream  *dns.ClientConfig
-	tcpClient *dns.Client
-	udpClient *dns.Client
+	ns               *Nameserver
+	domain           string
+	ttl              uint32
+	address          string
+	effectiveAddress string
+
+	servers          []*dns.Server
+	upstream         *dns.ClientConfig
+	upstreamMutex    sync.Mutex
+	upstreamChecked  time.Time
+	upstreamModified time.Time
```

Comment:
I reckon encapsulating these four plus the effectiveAddress in a separate struct would improve the code.